presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -561,6 +545,75 @@ private void invokeNoMoreSplitsIfNecessary()
         return COMPLETED_FUTURE;
     }
 
+    private Iterator<InternalHiveSplit> buildManifestFileIterator(Table table,
+            FileStatus symlinkFile,
+            FileSystem fileSystem,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+    {
+        // Strategy: List once, filter file listing based on manifest file. Complication is that not all files in the manifest are guaranteed
+        // to have a common root.
+
+        List<Iterator<LocatedFileStatus>> fileStatusIterators = new ArrayList<>();
+        List<Path> paths = new ArrayList<>();
+
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(fileSystem.open(symlinkFile.getPath()), StandardCharsets.UTF_8))) {
+            CharStreams.readLines(reader).stream()
+                    .map(Path::new)
+                    .forEach(paths::add);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_BAD_DATA, "Error parsing symlinks from: " + symlinkFile, e);
+        }
+
+        ImmutableSet<Path> parents = paths.stream()
+                .map(Path::getParent)
+                .distinct()
+                .collect(toImmutableSet());
+
+        for (Path parent : parents) {
+            try {
+                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, parent);
+                fileStatusIterators.add(new HiveFileIterator(table, parent, targetFilesystem, directoryLister, namenodeStats, IGNORED, true));
+            }
+            //There will be a single listing performed if all paths have a common parent. Worst case behavior is N listings for N files
+            catch (IOException e) {
+                throw new PrestoException(HIVE_FILE_NOT_FOUND, String.format("No Filesystem available for %s", parent), e);
+            }
+        }
+
+        //TODO: test with impersonation
+        return hdfsEnvironment.doAs(hdfsContext.getIdentity().getUser(), () -> fileStatusIterators.stream()
+                .flatMap(Streams::stream) // com.google.common.collect.Streams#stream(java.util.Iterator<T>)
+                .filter(foo -> paths.contains(foo.getPath()))
+                .map(status -> splitFactory.createInternalHiveSplit(status, OptionalInt.empty(), splittable, Optional.empty()))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .iterator());
+    }
+
+    private InputFormat<?, ?> mapSymlinkSerdeToInputFormat(Configuration configuration)
+    {
+        String serDe = table.getStorage().getStorageFormat().getSerDe();
```

Comment:
what if there are partitions? There is `io.prestosql.plugin.hive.metastore.Partition#getStorage` also

Reply:
Updated to use the partition schema.